dts/arm/atmel/sam4e16e.dtsi

```diff
@@ -12,6 +12,7 @@
 		flash-controller@400e0a00 {
 			flash0: flash@400000 {
 				reg = <0x00400000 DT_SIZE_K(1024)>;
+				erase-blocks = <&eefc 8 2048>, <&eefc 252 4096>;
```

Comment:
This is actually quite interesting. I wonder whether we could have some DT macro to directly obtain page size from given index.

Reply:
I was thinking the same, I would like to have the erase block phandles implemented as part of the `flash-controller.yaml` and have some proper macros for getting the pages if there is enough interest :)

Reply:
AT45 type devices could also use it. because they can have smaller pages at the beginning.
There are also STM32 devices that have different layouts, but this may be hard to do in DTS (I went through reading a lot of stm32 datasheets for this https://github.com/zephyrproject-rtos/zephyr/pull/36484).

Reply:
I like adding the information about the erase-block sizes to the dts. I am unsure if this should be part of the `flash-controller.yaml` as also spi devices might have this property. Does the erase-blocks need a phandle to the flash-controller ? I am also unsure if it is needed to incorporate the page-layout to avoid any overlap, the page layout is the most "generic" solution supporting erase-block sizes that "grow" and "reduce" in size, while the driver knows that it are only the first pages that are smaller. The driver could be smarter than just using the page_layout.

Reply:
To support also the stm32 devices the erase-blocks could be defined as `<offset blocksize blockcount>`, this could also simplify the verification of the area.

Reply:
@Laczen The phandle is required for the phandle-arrays, the phandle describes the contents of the array :) Regarding the `<offset blocksize blockcount>`, a flash area is contiguous, starting at offset 0, if there are multiple flash areas, they will get there own `erase-blocks`
```
		flash-controller@400e0a00 {
			flash0: flash@400000 {
				reg = <0x00400000 DT_SIZE_K(1024)>;
				erase-blocks = <&eefc 8 2048>, <&eefc 252 4096>;
			};
			flash1: flash@410000 {
				reg = <0x00410000 DT_SIZE_K(1024)>;
				erase-blocks = <&eefc 8 2048>, <&eefc 252 4096>;
			};
			...
```

Reply:
@de-nordic We could create a new bindings file in `bindings/mtd` named `mtd-page-layout.yaml` or similar that includes the `erase-blocks` phandle-array which can then be included by flash and at45 based devices

Reply:
@bjarki-trackunit It seems that the phandle-array is not really a good fit for describing the erase-blocks as it needs a phandle :). At the moment the stm32 devices that have a "discontiguous" flash are defined as one flash node (flash0), and the page_layout is one array. The `offset` would be used to allow a discontiguity to be described.

Reply:
@Laczen The phandle to the flash controller allows you to define the contents of the `erase-blocks` to match that controller.

In atmel-sam-flash-controller.yaml
```
  "#erase-block-cells":
    type: int
    const: 2

erase-block-cells:
  - pages-count
  - pages-size
```
which is why `erase-blocks = <&eefc 8 2048>, <&eefc 252 4096>;` becomes
```
#define ... IDX_0_pages_count 8
#define ... IDX_0_pages_size 2048
#define ... IDX_1_pages_count 252
#define ... IDX_1_pages_size 4096
```
The controller (phandle) defines the specific contents of the property. If the STM32 requires the offset as well, it can define it as:
```
  "#erase-block-cells":
    type: int
    const: 3

erase-block-cells:
  - offset
  - pages-count
  - pages-size
```
Which allows you to define
```
`erase-blocks = <&flash 0 8 2048>, <&flash 0x4000 252 4096>;`
```
while still being compatible with the pages_size and pages_count macro :)

I still believe you should define multiple flash areas if there are multiple flash areas, it feels like the offset is a workaround to me :)